File: app/src/main/java/com/zxcx/zhizhe/ui/card/cardBag/CardBagActivity.java
package com.zxcx.zhizhe.ui.card.cardBag;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.kingja.loadsir.core.LoadSir;
import com.zxcx.zhizhe.R;
import com.zxcx.zhizhe.loadCallback.CardBagLoadingCallback;
import com.zxcx.zhizhe.loadCallback.LoginTimeoutCallback;
import com.zxcx.zhizhe.loadCallback.NetworkErrorCallback;
import com.zxcx.zhizhe.mvpBase.RefreshMvpActivity;
import com.zxcx.zhizhe.ui.card.card.newCardDetails.CardDetailsActivity;
import com.zxcx.zhizhe.ui.card.cardBag.adapter.CardBagCardAdapter;
import com.zxcx.zhizhe.ui.card.cardBag.adapter.CardBagListAdapter;
import com.zxcx.zhizhe.utils.Constants;
import com.zxcx.zhizhe.utils.DateTimeUtils;
import com.zxcx.zhizhe.widget.CustomLoadMoreView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import in.srain.cube.views.ptr.PtrFrameLayout;

public class CardBagActivity extends RefreshMvpActivity<CardBagPresenter> implements CardBagContract.View,
        BaseQuickAdapter.RequestLoadMoreListener {


    @BindView(R.id.iv_toolbar_back)
    ImageView mIvToolbarBack;
    @BindView(R.id.toolbar_title)
    TextView mToolbarTitle;
    @BindView(R.id.iv_toolbar_right)
    ImageView mIvToolbarRight;
    @BindView(R.id.rv_card_bag_card)
    RecyclerView mRvCardBagCard;
    private CardBagCardAdapter mCardBagCardAdapter;
    private CardBagListAdapter mCardBagListAdapter;
    private CardBagCardItemDecoration mCardItemDecoration = new CardBagCardItemDecoration();
    private CardBagListItemDecoration mListItemDecoration = new CardBagListItemDecoration();
    private boolean isCard = true;
    private int showFistItem;
    private LinearLayoutManager mCardBagCardManager;
    private int mId;
    private int page = 0;
    private String mName;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setContentView(R.layout.activity_card_bag);
        super.onCreate(savedInstanceState);
        ButterKnife.bind(this);
        initData();
        initRecyclerView();
        mIvToolbarRight.setVisibility(View.VISIBLE);
        mIvToolbarRight.setImageResource(R.drawable.iv_card_bag_list);
        onRefresh();

        initLoadSir();
    }

    @Override
    protected CardBagPresenter createPresenter() {
        return new CardBagPresenter(this);
    }

    public void initLoadSir() {
        LoadSir loadSir = new LoadSir.Builder()
                .addCallback(new CardBagLoadingCallback())
                .addCallback(new NetworkErrorCallback())
                .addCallback(new LoginTimeoutCallback())
                .setDefaultCallback(CardBagLoadingCallback.class)
                .build();
        loadService = loadSir.register(this, this);
    }

    @Override
    public void onReload(View v) {
        loadService.showCallback(CardBagLoadingCallback.class);
        onRefresh();
    }

    @Override
    public void onRefreshBegin(PtrFrameLayout frame) {
        onRefresh();
    }

    public void onRefresh() {
        page = 0;
        getCardBagCardList();
    }

    @Override
    public void onLoadMoreRequested() {
        getCardBagCardList();
    }

    @Override
    public void getDataSuccess(List<CardBagBean> list) {
        loadService.showSuccess();
        mRefreshLayout.refreshComplete();
        if (page == 0) {
            mCardBagCardAdapter.setNewData(list);
            mCardBagListAdapter.setNewData(list);
        } else {
            mCardBagCardAdapter.addData(list);
            mCardBagListAdapter.addData(list);
        }
        page++;
        if (list.size() < Constants.PAGE_SIZE) {
            mCardBagCardAdapter.loadMoreEnd(false);
            mCardBagListAdapter.loadMoreEnd(false);
        } else {
            mCardBagCardAdapter.loadMoreComplete();
            mCardBagListAdapter.loadMoreComplete();
            mCardBagCardAdapter.setEnableLoadMore(false);
            mCardBagListAdapter.setEnableLoadMore(false);
            mCardBagCardAdapter.setEnableLoadMore(true);
            mCardBagListAdapter.setEnableLoadMore(true);
        }
    }

    @Override
    public void toastFail(String msg) {
        mRefreshLayout.refreshComplete();
        super.toastFail(msg);
        if (page == 0) {
            loadService.showCallback(NetworkErrorCallback.class);
        }
    }

    @OnClick(R.id.iv_toolbar_right)
    public void onMIvToolbarRightClicked() {
        if (isCard) {
            isCard = !isCard;
            showFistItem = mCardBagCardManager.findFirstVisibleItemPosition();
            mRvCardBagCard.setAdapter(mCardBagListAdapter);
            mRvCardBagCard.removeItemDecoration(mCardItemDecoration);
            mRvCardBagCard.addItemDecoration(mListItemDecoration);
            mIvToolbarRight.setImageResource(R.drawable.iv_card_bag_card);
            mCardBagCardManager.scrollToPosition(showFistItem);
        } else {
            isCard = !isCard;
            showFistItem = mCardBagCardManager.findFirstVisibleItemPosition();
            mRvCardBagCard.setAdapter(mCardBagCardAdapter);
            mRvCardBagCard.removeItemDecoration(mListItemDecoration);
            mRvCardBagCard.addItemDecoration(mCardItemDecoration);
            mIvToolbarRight.setImageResource(R.drawable.iv_card_bag_list);
            mCardBagCardManager.scrollToPosition(showFistItem);
        }
    }

    private void getCardBagCardList() {
        mPresenter.getCardBagCardList(mId, page, Constants.PAGE_SIZE);
    }

    private void initData() {
        mId = getIntent().getIntExtra("id", 0);
        mName = getIntent().getStringExtra("name");
        initToolBar(mName);
    }

    private void initRecyclerView() {
        mCardBagCardManager = new LinearLayoutManager(mActivity,LinearLayoutManager.VERTICAL,false);
        mCardBagCardAdapter = new CardBagCardAdapter(new ArrayList<>());
        mCardBagCardAdapter.setLoadMoreView(new CustomLoadMoreView());
        mCardBagCardAdapter.setOnLoadMoreListener(this, mRvCardBagCard);
        mCardBagCardAdapter.setOnItemClickListener(new CardItemClickListener(this));
        mRvCardBagCard.setLayoutManager(mCardBagCardManager);
        mRvCardBagCard.setAdapter(mCardBagCardAdapter);
        mRvCardBagCard.addItemDecoration(mCardItemDecoration);
        mCardBagCardAdapter.setOnItemClickListener(new CardItemClickListener(this));

        mCardBagListAdapter = new CardBagListAdapter(new ArrayList<>());
        mCardBagListAdapter.setLoadMoreView(new CustomLoadMoreView());
        mCardBagListAdapter.setOnLoadMoreListener(this, mRvCardBagCard);
        mCardBagListAdapter.setOnItemClickListener(new CardItemClickListener(this));

        View view = LayoutInflater.from(mActivity).inflate(R.layout.layout_no_data, null);
        mCardBagCardAdapter.setEmptyView(view);
    }

    class CardItemClickListener implements BaseQuickAdapter.OnItemClickListener {

        private Context mContext;

        public CardItemClickListener(Context context) {
            mContext = context;
        }

        @Override
        public void onItemClick(BaseQuickAdapter adapter, View view, int position) {
            CardBagBean bean = (CardBagBean) adapter.getData().get(position);
            Intent intent = new Intent(mContext, CardDetailsActivity.class);
            intent.putExtra("id", bean.getId());
            intent.putExtra("name", bean.getName());
            intent.putExtra("imageUrl", bean.getImageUrl());
            intent.putExtra("date", DateTimeUtils.getDateString(bean.getDate()));
            intent.putExtra("author", bean.getAuthor());
            mContext.startActivity(intent);
        }
    }

}

File: app/src/main/java/com/zxcx/zhizhe/ui/card/cardBag/CardBagActivity.kt
package com.zxcx.zhizhe.ui.card.cardBag

import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import butterknife.ButterKnife
import com.chad.library.adapter.base.BaseQuickAdapter
import com.kingja.loadsir.core.LoadSir
import com.scwang.smartrefresh.layout.api.RefreshLayout
import com.zxcx.zhizhe.R
import com.zxcx.zhizhe.loadCallback.LoadingCallback
import com.zxcx.zhizhe.loadCallback.LoginTimeoutCallback
import com.zxcx.zhizhe.loadCallback.NetworkErrorCallback
import com.zxcx.zhizhe.mvpBase.RefreshMvpActivity
import com.zxcx.zhizhe.ui.article.articleDetails.ArticleDetailsActivity
import com.zxcx.zhizhe.ui.article.attention.SubjectBean
import com.zxcx.zhizhe.ui.article.attention.SubjectOnClickListener
import com.zxcx.zhizhe.ui.card.hot.CardAdapter
import com.zxcx.zhizhe.ui.card.hot.CardBean
import com.zxcx.zhizhe.ui.classify.subject.SubjectArticleActivity
import com.zxcx.zhizhe.utils.Constants
import com.zxcx.zhizhe.utils.DateTimeUtils
import com.zxcx.zhizhe.widget.CustomLoadMoreView
import com.zxcx.zhizhe.widget.EmptyView
import kotlinx.android.synthetic.main.activity_card_bag.*
import java.util.*

class CardBagActivity : RefreshMvpActivity<CardBagPresenter>(), CardBagContract.View,
		BaseQuickAdapter.RequestLoadMoreListener, BaseQuickAdapter.OnItemClickListener,
		SubjectOnClickListener {

	private lateinit var mAdapter: CardAdapter
	private var mId: Int = 0
	private var page = 0
	private var name = ""

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_card_bag)
		ButterKnife.bind(this)
		initData()
		initView()
		onRefresh()

		initLoadSir()
	}

	override fun createPresenter(): CardBagPresenter {
		return CardBagPresenter(this)
	}

	private fun initLoadSir() {
		val loadSir = LoadSir.Builder()
				.addCallback(LoadingCallback())
				.addCallback(NetworkErrorCallback())
				.addCallback(LoginTimeoutCallback())
				.setDefaultCallback(LoadingCallback::class.java)
				.build()
		loadService = loadSir.register(mRefreshLayout, this)
	}

	override fun onReload(v: View) {
		loadService.showCallback(LoadingCallback::class.java)
		onRefresh()
	}

	override fun onRefresh(refreshLayout: RefreshLayout) {
		onRefresh()
	}

	fun onRefresh() {
		page = 0
		getCardBagCardList()
	}

	override fun onLoadMoreRequested() {
		getCardBagCardList()
	}

	override fun getDataSuccess(list: List<CardBean>) {
		loadService.showSuccess()
		mRefreshLayout.finishRefresh()
		if (page == 0) {
			mAdapter.setNewData(list)
		} else {
			mAdapter.addData(list)
		}
		page++
		if (list.size < Constants.PAGE_SIZE) {
			mAdapter.loadMoreEnd(false)
		} else {
			mAdapter.loadMoreComplete()
			mAdapter.setEnableLoadMore(false)
			mAdapter.setEnableLoadMore(true)
		}
	}

	override fun toastFail(msg: String) {
		if (page == 0) {
			loadService.showCallback(NetworkErrorCallback::class.java)
		}
		super.toastFail(msg)
	}

	private fun getCardBagCardList() {
		mPresenter.getCardBagCardList(mId, page, Constants.PAGE_SIZE)
	}

	private fun initData() {
		mId = intent.getIntExtra("id", 0)
		name = intent.getStringExtra("name")
		tv_header_title.text = name
	}

	private fun initView() {
		mAdapter = CardAdapter(ArrayList())
		mAdapter.setLoadMoreView(CustomLoadMoreView())
		mAdapter.setOnLoadMoreListener(this, rv_card_bag_card)
		mAdapter.onItemClickListener = this
		val view = EmptyView.getEmptyView(mActivity, "暂无内容", R.drawable.no_data)
		mAdapter.emptyView = view

		rv_card_bag_card.layoutManager = LinearLayoutManager(mActivity, LinearLayoutManager.VERTICAL, false)
		rv_card_bag_card.adapter = mAdapter
	}

	override fun setListener() {
		iv_common_close.setOnClickListener {
			onBackPressed()
		}
	}

	override fun onItemClick(adapter: BaseQuickAdapter<*, *>, view: View, position: Int) {
		/*val CardBean = adapter.data[position] as CardBean
		if (CardBean.itemType == CardBean.TYPE_CARD) {
			val bean = CardBean.cardBean
			val bundle = ActivityOptionsCompat.makeSceneTransitionAnimation(this,
					Pair.create(view.findViewById(R.id.iv_item_card_icon), "cardImage"),
					Pair.create(view.findViewById(R.id.tv_item_card_title), "cardTitle"),
					Pair.create(view.findViewById(R.id.tv_item_card_category), "cardBag")).toBundle()
			val intent = Intent(this, ArticleDetailsActivity::class.java)
			intent.putExtra("id", bean.id)
			intent.putExtra("name", bean.name)
			intent.putExtra("imageUrl", bean.imageUrl)
			intent.putExtra("date", DateTimeUtils.getDateString(bean.date))
			intent.putExtra("authorName", bean.authorName)
			startActivity(intent, bundle)
		}*/
	}

	override fun articleOnClick(bean: CardBean) {
		val intent = Intent(this, ArticleDetailsActivity::class.java)
		intent.putExtra("id", bean.id)
		intent.putExtra("name", bean.name)
		intent.putExtra("imageUrl", bean.imageUrl)
		intent.putExtra("date", DateTimeUtils.getDateString(bean.date))
		intent.putExtra("authorName", bean.authorName)
		startActivity(intent)
	}

	override fun subjectOnClick(bean: SubjectBean) {
		val intent = Intent(this, SubjectArticleActivity::class.java)
		intent.putExtra("id", bean.id)
		intent.putExtra("name", bean.name)
		startActivity(intent)
	}

}
